src/main/java/com/aws/greengrass/lifecyclemanager/Kernel.java

```diff
@@ -196,24 +199,19 @@ public Kernel launch() {
         switch (deploymentStageAtLaunch) {
             case BOOTSTRAP:
                 logger.atInfo().kv("deploymentStage", deploymentStageAtLaunch).log("Resume deployment");
-                int exitCode;
                 try {
-                    exitCode = bootstrapManager.executeAllBootstrapTasksSequentially(
-                            deploymentDirectoryManager.getBootstrapTaskFilePath());
-                    if (!bootstrapManager.hasNext()) {
-                        logger.atInfo().log("Completed all bootstrap tasks. Continue to activate deployment changes");
-                    }
-                    // If exitCode is 0, which happens when all bootstrap tasks are completed, restart in new launch
-                    // directories and verify handover is complete. As a result, exit code 0 is treated as 100 here.
-                    logger.atInfo().log((exitCode == REQUEST_REBOOT ? "device reboot" : "Nucleus restart")
-                            + " requested to complete bootstrap task");
-
-                    shutdown(30, exitCode == REQUEST_REBOOT ? REQUEST_REBOOT : REQUEST_RESTART);
+                    Path bootstrapTaskFilePath = deploymentDirectoryManager.getBootstrapTaskFilePath();
+                    executeBootstrapTasksAndShutdown(bootstrapManager, bootstrapTaskFilePath);
                 } catch (ServiceUpdateException | IOException e) {
                     logger.atError().log("Deployment bootstrap failed", e);
                     try {
+                        // Bootstrapping for target deployment failed, so check if bootstrap-on-rollback is needed
+                        boolean bootstrapOnRollbackRequired = kernelAlts.prepareBootstrapOnRollbackIfNeeded(
+                                this.context, deploymentDirectoryManager, bootstrapManager);
+                        // Save deployment error information
                         Deployment deployment = deploymentDirectoryManager.readDeploymentMetadata();
-                        deployment.setDeploymentStage(DeploymentStage.KERNEL_ROLLBACK);
+                        deployment.setDeploymentStage(
+                                bootstrapOnRollbackRequired ? ROLLBACK_BOOTSTRAP : KERNEL_ROLLBACK);
                         Pair<List<String>, List<String>> errorReport =
                                 DeploymentErrorCodeUtils.generateErrorReportFromExceptionStack(e);
                         deployment.setErrorStack(errorReport.getLeft());
```

Comment:
I think we should modify the log in the catch section for this try block. Is it possible for the try block to throw IOException besides not being able to read deployment metadata?

Reply:
`kernelAlts.prepareBootstrapOnRollbackIfNeeded()` doesn't throw IOException.

Only sources of IOException are `deploymentDirectoryManager.readDeploymentMetadata()` and `deploymentDirectoryManager.writeDeploymentMetadata()`.

We could change the log text from "Could not read" to "Could not read/write"